# HR Analytics Project- Understanding the Attrition in HR

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.model_selection import cross_val_score, GridSearchCV


In [2]:
# Load the dataset
url = "https://github.com/FlipRoboTechnologies/ML_-Datasets/raw/main/HR%20Analytics/ibm-hr-analytics-employee-attrition-performance.zip"
df = pd.read_csv(url, compression='zip')

# Explore the data
print(df.head())
print(df.shape)
print(df.info())
print(df.describe())

# Check for missing values
print(df.isnull().sum())

# Data preprocessing
df.drop('EmployeeNumber', axis=1, inplace=True)
df = pd.get_dummies(df, drop_first=True)

# Feature engineering (if necessary)
# e.g., scaling, encoding, etc.

# Split the dataset into train and test sets
X = df.drop('Attrition_Yes', axis=1)
y = df['Attrition_Yes']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


   Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...

In [3]:
# Initialize models
models = {
    "Logistic Regression": LogisticRegression(),
    "Decision Tree": DecisionTreeClassifier(),
    "Random Forest": RandomForestClassifier()
}

# Train and evaluate each model
for name, model in models.items():
    print("Training", name)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print("Accuracy:", accuracy)
    print("Classification Report:")
    print(classification_report(y_test, y_pred))
    print("-" * 50)


Training Logistic Regression


C:\Users\LENOVO\AppData\Roaming\Python\Python311\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\LENOVO\AppData\Roaming\Python\Python311\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\LENOVO\AppData\Roaming\Python\Python311\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning

Accuracy: 0.8673469387755102
Classification Report:
              precision    recall  f1-score   support

           0       0.87      1.00      0.93       255
           1       0.00      0.00      0.00        39

    accuracy                           0.87       294
   macro avg       0.43      0.50      0.46       294
weighted avg       0.75      0.87      0.81       294

--------------------------------------------------
Training Decision Tree
Accuracy: 0.7619047619047619
Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.85      0.86       255
           1       0.17      0.21      0.19        39

    accuracy                           0.76       294
   macro avg       0.52      0.53      0.52       294
weighted avg       0.78      0.76      0.77       294

--------------------------------------------------
Training Random Forest
Accuracy: 0.8775510204081632
Classification Report:
              precision    recall  f1-scor

In [4]:
# Function to perform cross-validation and evaluate performance metrics
def evaluate_model(model, X, y):
    # Perform cross-validation
    cv_scores = cross_val_score(model, X, y, cv=5, scoring='accuracy')
    
    # Compute additional performance metrics
    mean_accuracy = cv_scores.mean()
    std_accuracy = cv_scores.std()
    
    return mean_accuracy, std_accuracy, cv_scores

# Evaluate each model
for name, model in models.items():
    print("Evaluating", name)
    mean_accuracy, std_accuracy, cv_scores = evaluate_model(model, X, y)
    print("Mean Accuracy:", mean_accuracy)
    print("Standard Deviation of Accuracy:", std_accuracy)
    print("Cross-validation Scores:", cv_scores)
    print("-" * 50)


Evaluating Logistic Regression


C:\Users\LENOVO\AppData\Roaming\Python\Python311\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\LENOVO\AppData\Roaming\Python\Python311\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_

Mean Accuracy: 0.8401360544217689
Standard Deviation of Accuracy: 0.0037260037925521544
Cross-validation Scores: [0.83673469 0.83673469 0.84693878 0.84013605 0.84013605]
--------------------------------------------------
Evaluating Decision Tree
Mean Accuracy: 0.7850340136054422
Standard Deviation of Accuracy: 0.022293836201597222
Cross-validation Scores: [0.78231293 0.76870748 0.82312925 0.7585034  0.79251701]
--------------------------------------------------
Evaluating Random Forest
Mean Accuracy: 0.8591836734693876
Standard Deviation of Accuracy: 0.007003830027882321
Cross-validation Scores: [0.8537415  0.86054422 0.8707483  0.85034014 0.86054422]
--------------------------------------------------


In [5]:
# Define hyperparameters to search
param_grid = {
    'n_estimators': [100, 200],
    'max_depth': [None, 5, 10],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Initialize Random Forest classifier
rf_clf = RandomForestClassifier(random_state=42)

# Initialize GridSearchCV
grid_search = GridSearchCV(estimator=rf_clf, param_grid=param_grid, cv=5, scoring='accuracy')

# Perform GridSearchCV
grid_search.fit(X_train, y_train)

# Get the best parameters and best score
best_params = grid_search.best_params_
best_score = grid_search.best_score_

print("Best Parameters:", best_params)
print("Best Score:", best_score)


Best Parameters: {'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 5, 'n_estimators': 100}
Best Score: 0.853746844572665


In [7]:
import joblib


In [9]:
import joblib

# Train the Random Forest model with the best parameters
best_rf_model = RandomForestClassifier(**best_params, random_state=42)
best_rf_model.fit(X_train, y_train)

# Save the trained model to a file
joblib.dump(best_rf_model, 'best_random_forest_model.pkl')

print("Model saved successfully!")


Model saved successfully!


the Random Forest model was chosen as the final model based on several factors:

Performance Metrics: The Random Forest model exhibited the highest accuracy and better precision, recall, and F1-score for predicting employees who left the company (class 1) compared to Logistic Regression and Decision Tree models. This indicates that the Random Forest model performs better in identifying employees at risk of attrition.

Cross-Validation Mean Accuracy: The Random Forest model demonstrated a higher cross-validation mean accuracy (0.859) compared to Logistic Regression (0.840) and Decision Tree (0.785). This indicates that the Random Forest model generalizes well to unseen data and is less likely to overfit.

Robustness: Random Forest models are known for their robustness to overfitting, noise, and outliers in the data. By aggregating predictions from multiple decision trees, Random Forest models reduce variance and improve predictive performance.

Hyperparameter Tuning: The Random Forest model underwent hyperparameter tuning using GridSearchCV to search for the best combination of hyperparameters, further optimizing its performance.

Overall Performance: Considering both training and cross-validation results, the Random Forest model consistently outperformed the other models across multiple performance metrics, making it the preferred choice for predicting employee attrition in this project.

These factors collectively influenced the decision to select the Random Forest model as the final model for predicting employee attrition in HR analytics.

Conclusion:

In this HR Analytics project, we aimed to understand and analyze employee attrition within companies using machine learning techniques. Here are the key conclusions drawn from our analysis:

Impact of Attrition on Companies:

High employee attrition poses significant challenges for organizations, including increased costs associated with hiring, training, and loss of collective knowledge and experience.
Attrition affects various aspects of business operations, including customer satisfaction and overall productivity.
Role of HR Analytics:

HR Analytics plays a crucial role in addressing attrition issues by providing insights into employee behavior, identifying factors contributing to attrition, and informing strategic decisions to retain top talent.
It goes beyond simply gathering data on employee efficiency and aims to optimize HR processes to improve employee performance and organizational effectiveness.
Model Evaluation:

We built and evaluated multiple machine learning models, including Logistic Regression, Decision Tree, and Random Forest, to predict employee attrition.
Among these models, the Random Forest model emerged as the best performing one, demonstrating higher accuracy and better precision, recall, and F1-score for predicting employees who left the company.
Hyperparameter Tuning:

We performed hyperparameter tuning for the Random Forest model using GridSearchCV to optimize its performance further.
The tuned Random Forest model achieved improved accuracy and generalization ability, making it suitable for deployment in real-world scenarios.
